File: algorithms/Python/smallestDiff.py
# O(nlog(n) + mlog(m)) Time | O(1) Space
def smallestDiff(arr1, arr2):
    arr1.sort()
    arr2.sort()
    idx1 = 0
    idx2 = 0
    smallest = float("inf")
    current = float("inf")
    bestPair = []
    while idx1 < len(arr1) and idx2 < len(arr2):
        num1 = arr1[idx1]
        num2 = arr2[idx2]
        if num1 < num2:
            current = num2 - num1
            idx1 += 1
        elif num2 < num1:
            current = num1 - num2
            idx2 += 1
        else: # they're equal, their diff is 0
            return [num1, num2]
        if current < smallest:
            smallest = current
            bestPair = [num1, num2]
    return bestPair
    

File: algorithms/Javascript/smallestDiff.js
// Function takes two arrays of integers. Find bestPairs between the arrays
// where the absolute difference is closest to zero

// O(nlog(n) + mlog(n)) Time | O(1) Space
function smallestDiff(array1, array2) {
  array1.sort((a,b) => a - b);
  array2.sort((a,b) => a - b);
  let idx1 = 0;
  let idx2 = 0;
  let smallest = Infinity; // used for comparing first pair, which is gauranteed smaller
  let current = Infinity; // need arbitrary num to kickstart algo, but can't use null
  let bestPair = [];
  while (idx1 < array1.length && idx2 < array2.length) {
    let num1 = array1[idx1];
    let num2 = array2[idx2];
    if (num1 < num2) {
      current = num2 - num1;
      idx1++;
    } else if (num2 < num1) {
      current = num1 - num2;
      idx2++;
    } else { // they're equal, abs diff equates to 0
      return [num1, num2];
    }
    if (current < smallest) {
      smallest = current;
      bestPair = [num1, num2];
    }
  }
  return bestPair;
}